src/QsCompiler/Transformations/CallGraph/SimpleCallGraph.cs

```diff
@@ -0,0 +1,470 @@
+
+        // Constructors
+
+        /// <summary>
+        /// Constructs a call graph from a compilation. The optional trim argument may be
+        /// used to specify the call graph to be trimmed to only include callables that
+        /// entry points are dependent on.
+        /// Throws ArgumentNullException if compilation argument is null.
+        /// </summary>
+        public SimpleCallGraph(QsCompilation compilation, bool trim = false)
+        {
+            if (compilation is null)
+            {
+                throw new ArgumentNullException(nameof(compilation));
+            }
+
+            if (trim)
+            {
+                BuildCallGraph.PopulateTrimmedGraph(this.graphBuilder, compilation);
+            }
+            else
+            {
+                BuildCallGraph.PopulateSimpleGraph(this.graphBuilder, compilation);
+            }
+        }
+
+        /// <summary>
+        /// Constructs a call graph from callables.
+        /// Throws an ArgumentNullException if the argument is null or any of the callables are null.
+        /// </summary>
+        public SimpleCallGraph(IEnumerable<QsCallable> callables)
+        {
+            if (callables is null || callables.Any(x => x is null))
+            {
+                throw new ArgumentNullException(nameof(callables));
+            }
+
+            BuildCallGraph.PopulateSimpleGraph(this.graphBuilder, callables);
+        }
+
+        // Member Methods
+
+        /// <summary>
+        /// Returns the children nodes of a given node. Each key in the returned lookup is a child
+        /// node of the given node. Each value in the lookup is an edge connecting the given node to
+        /// the child node represented by the associated key.
+        /// Returns an empty ILookup if the node was found with no dependencies or was not found in
+        /// the graph.
+        /// Throws ArgumentNullException if argument is null.
+        /// </summary>
+        public ILookup<SimpleCallGraphNode, SimpleCallGraphEdge> GetDirectDependencies(SimpleCallGraphNode node) => this.graphBuilder.GetDirectDependencies(node);
+
+        /// <summary>
+        /// Given a call graph edges, finds all cycles and determines if each is valid.
+        /// Invalid cycles are those that cause type parameters to be mapped to
+        /// other type parameters of the same callable (constricting resolutions)
+        /// or to a type containing a nested reference to the same type parameter,
+        /// i.e Foo.A -> Foo.A[].
+        /// Returns an enumerable of tuples for each edge of each invalid cycle found,
+        /// each tuple containing a diagnostic and the callable name where the diagnostic
+        /// should be placed.
+        /// </summary>
+        public IEnumerable<Tuple<QsCompilerDiagnostic, QsQualifiedName>> VerifyAllCycles()
+        {
+            var diagnostics = new List<Tuple<QsCompilerDiagnostic, QsQualifiedName>>();
```

Comment:
I think in general in this code base it is standard to use System.Tuple because it has better interoperability with F#. This code in particular does not interact with F#, but as a matter of consistency I'd like to keep it as a System.Tuple unless there is a good reason to use the value tuple.

Reply:
Yes, F# tuples use `System.Tuple`, but the standard tuple in C# is `System.ValueTuple`, which also has nice built-in syntax. That is used throughout the compiler codebase as well as `System.Tuple` when needed for F# interop.

Reply:
Okay so in your experience with the code base System.Tuple is only used in the C# when necessary for F# interop? Alright then I can go for that. Value tuple _does_ have an awfully nice syntax. 😊

Reply:
changed